lib/puppet/parser/functions/hiera.rb

```diff
@@ -1,33 +1,88 @@
-  1. Using 1 to 3 arguments where the arguments are:
-     'key'      [String] Required
-           The key to lookup.
-     'default`  [Any] Optional
-           A value to return when there's no match for `key`. Optional
-     `override` [Any] Optional
-           An argument in the third position, providing a data source
-           to consult for matching values, even if it would not ordinarily be
-           part of the matched hierarchy. If Hiera doesn't find a matching key
-           in the named override data source, it will continue to search through the
-           rest of the hierarchy.
-
-  2. Using a 'key' and an optional 'override' parameter like in #1 but with a block to
-     provide the default value. The block is called with one parameter (the key) and
-     should return the value.
-
-  3. Like #1 but with all arguments passed in an array.
-
-  More thorough examples of `hiera` are available at:  
-  <http://docs.puppetlabs.com/hiera/1/puppet.html#hiera-lookup-functions>
-  ") do |*args|
+  newfunction(
+    :hiera,
+    :type => :rvalue,
+    :arity => -2,
+    :doc => <<-DOC
+Performs a standard priority lookup of the hierarchy and returns the most specific value
+for a given key. The returned value can be any type of data, depending on the values'
+types and the number of results.
+
+The function takes up to three arguments, in this order:
+
+1. A string key that Hiera searches for in the hierarchy. **Required**.
+2. An optional default value to return if Hiera doesn't find anything matching the key.
+  * If this argument isn't provided and this function results in a lookup failure, Puppet
+fails with a compilation error.
+3. The optional name of an arbitrary
+[hierarchy level](https://docs.puppetlabs.com/hiera/latest/hierarchy.html) to insert at the
+top of the hierarchy. This lets you temporarily modify the hierarchy for a single lookup.
+  * If Hiera doesn't find a matching key in the overriding hierarchy level, it continues
+searching the rest of the hierarchy.
+
+The `hiera` function does **not** find all matches throughout a hierarchy, instead
+returining the first specific value starting at the top of the hierarchy. To search
+throughout a hierarchy, use the `hiera_array` or `hiera_hash` functions.
+
+@example Using `hiera`
+
+~~~ puppet
+# Assuming hiera.yaml
+# :hierarchy:
+#   - web01.example.com
+#   - common
+
+# Assuming web01.example.com.yaml:
+# users: 
+#   - "Amy Barry"
+#   - "Carrie Douglas"
+
+# Assuming common.yaml:
+users:
+  admins:
+    - "Edith Franklin"
+    - "Ginny Hamilton"
+  regular:
+    - "Iris Jackson"
+    - "Kelly Lambert"
+
+$users = hiera('users', undef)
+
+# $users contains {admins => [Edith Franklin, Ginny Hamilton],
+#                 regular => [Iris Jackson, Kelly Lambert]}
+~~~
+
+Optionally, you can use a
+[lambda](https://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html) instead of
+an argument to generate the default value.
```

Comment:
It looks like the lambda takes one parameter, which is the key you provided as the function's first argument? We should say so here.